File: sdks/aihub-node/examples/langchain-demo.ts
/**
 * LangChain + AI Hub Integration Demo
 *
 * Demonstrates using the AI Hub LangchainCallbackHandler with LangChain.js.
 *
 * Prerequisites:
 *   npm install @langchain/core @langchain/openai
 *
 * Usage:
 *   1. Set AIHUB_API_KEY=aihub_sk_xxx
 *   2. npx tsx examples/langchain-demo.ts
 */
import { config } from 'dotenv';
config({ override: true });

import { ChatOpenAI } from '@langchain/openai';
import { LangchainCallbackHandler } from '../src/langchain';

const GATEWAY_URL = process.env['AIHUB_GATEWAY_URL'] || 'http://localhost:8787/v1';
const API_KEY = process.env['AIHUB_API_KEY'] || '';

// Create the AI Hub LangChain callback handler
const handler = new LangchainCallbackHandler({
  apiKey: API_KEY,
  baseURL: GATEWAY_URL,
  metadata: { demo: true, source: 'langchain-demo' },
});

// Initialize LangChain LLM with AI Hub as the backend
const llm = new ChatOpenAI({
  model: 'deepseek-chat',
  temperature: 0.7,
  configuration: {
    baseURL: GATEWAY_URL,
    apiKey: API_KEY,
  },
  callbacks: [handler],
});

async function main() {
  console.log('=== LangChain + AI Hub Demo ===\n');

  // 1. Basic invoke
  console.log('1. Basic invoke:');
  const response = await llm.invoke('用一句话介绍AI Hub。');
  console.log('   Response:', response.content);
  console.log('   Tokens:', response.response_metadata?.tokenUsage);

  // 2. Batch invoke
  console.log('\n2. Batch invoke:');
  const responses = await llm.batch([
    '什么是LLM网关？',
    'Agent控制平面有什么作用？',
  ]);
  responses.forEach((r, i) => {
    console.log(`   [${i + 1}] ${r.content}`);
  });

  // 3. Streaming
  console.log('\n3. Streaming:');
  const stream = await llm.stream('写一首三行诗。');
  process.stdout.write('   ');
  for await (const chunk of stream) {
    process.stdout.write(chunk.content);
  }

  // Flush any remaining events
  await handler.flush();

  console.log('\n\n4. Buffered events:', handler.getEvents().length);

  console.log('\n✅ Demo complete! Check your AI Hub Dashboard for the recorded events.');
}

main().catch(console.error);

File: sdks/aihub-python/examples/langchain_demo.py
"""
LangChain + AI Hub Integration Demo

Demonstrates using the AI Hub LangchainCallbackHandler to trace
LangChain LLM calls, chains, and agent operations.

Supports TWO modes:
    Mode A: Callback Handler (auto-tracing, zero code change)
    Mode B: PortkeyLLM / ChatPortkey wrapper (direct replacement)

Prerequisites:
    pip install aihub[langchain_callback] langchain langchain-openai

Usage:
    1. Set AIHUB_API_KEY=aihub_sk_xxx
    2. python examples/langchain_demo.py
"""
import os
from dotenv import load_dotenv

load_dotenv()

GATEWAY_URL = os.environ.get("AIHUB_GATEWAY_URL", "http://localhost:8787/v1")
API_KEY = os.environ.get("AIHUB_API_KEY", "")

# ═══════════════════════════════════════════════════════════════
# Mode A: Callback Handler (auto-tracing)
# ═══════════════════════════════════════════════════════════════
def demo_callback_mode():
    print("=" * 50)
    print("Mode A: Callback Handler")
    print("=" * 50)

    from langchain_openai import ChatOpenAI
    from portkey_ai.langchain import LangchainCallbackHandler

    # Create the AI Hub callback
    handler = LangchainCallbackHandler(
        api_key=API_KEY,
        base_url=GATEWAY_URL,
        metadata={"demo": "langchain_callback", "mode": "A"},
    )

    # Use standard LangChain LLM with callback
    llm = ChatOpenAI(
        model="deepseek-chat",
        base_url=GATEWAY_URL,
        api_key=API_KEY,
        temperature=0.7,
        callbacks=[handler],
    )

    # 1. Single invoke
    print("\n1. Single invoke:")
    response = llm.invoke("用一句话介绍 AI Hub。")
    print(f"   {response.content}")

    # 2. Batch
    print("\n2. Batch invoke:")
    responses = llm.batch([
        "什么是 LLM 网关？",
        "Agent 控制平面的核心功能是什么？",
    ])
    for i, r in enumerate(responses, 1):
        print(f"   [{i}] {r.content[:80]}...")

    # 3. Streaming
    print("\n3. Streaming:")
    for chunk in llm.stream("写一个三行诗。"):
        print(chunk.content, end="", flush=True)
    print()

    print("\n✅ Mode A complete!")

# ═══════════════════════════════════════════════════════════════
# Mode B: ChatPortkey Wrapper (direct replacement)
# ═══════════════════════════════════════════════════════════════
def demo_wrapper_mode():
    print("\n" + "=" * 50)
    print("Mode B: ChatPortkey Wrapper")
    print("=" * 50)

    from portkey_ai.llms.langchain.chat import ChatPortkey

    # ChatPortkey directly wraps the LangChain ChatModel
    # No callback needed — auto-traces by design
    llm = ChatPortkey(
        api_key=API_KEY,
        base_url=GATEWAY_URL,
        model="deepseek-chat",
        temperature=0.7,
    )

    response = llm.invoke("用一句话介绍 AI Hub。")
    print(f"\nResponse: {response.content}")
    print(f"Model: {response.response_metadata.get('model', 'unknown')}")

    print("\n✅ Mode B complete!")
    print("\n" + "=" * 50)
    print("Both modes completed successfully!")
    print("Check your AI Hub Dashboard for the recorded events.")
    print("=" * 50)


if __name__ == "__main__":
    demo_callback_mode()
    demo_wrapper_mode()
